LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/Utils/TwinCollectionExtensions.cs

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#nullable enable
+
+namespace LoRaTools.Utils
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics.CodeAnalysis;
+    using System.Globalization;
+    using LoRaWan;
+    using Microsoft.Azure.Devices.Shared;
+    using Microsoft.Extensions.Logging;
+    using Newtonsoft.Json;
+    using Newtonsoft.Json.Linq;
+
+    public static class TwinCollectionExtensions
+    {
+        private static readonly Dictionary<Type, Func<object, object>> customConverters = new Dictionary<Type, Func<object, object>>
+        {
+            [typeof(StationEui)] = (s) => StationEui.Parse(s.ToString()),
+            [typeof(DevNonce)] = (s) => new DevNonce(Convert.ToUInt16(s, CultureInfo.InvariantCulture)),
+            [typeof(DevAddr)] = (s) => DevAddr.Parse(s.ToString()),
+        };
+
+        public static T? SafeRead<T>(this TwinCollection twinCollection, string property, T? defaultValue = default, ILogger? logger = null)
+            => twinCollection.TryRead<T>(property, logger, out var someT) ? someT : defaultValue;
+
+        public static string ReadRequiredString(this TwinCollection twinCollection, string property, ILogger? logger = null) =>
+                twinCollection.TryRead<string>(property, logger, out var someString) && !string.IsNullOrEmpty(someString)
+                                ? someString
+                                : throw new InvalidOperationException($"Property '{property}' does not exist or is empty.");
+
+        public static bool TryRead<T>(this TwinCollection twinCollection, string property, ILogger? logger, [NotNullWhen(true)] out T? value)
+        {
+            _ = twinCollection ?? throw new ArgumentNullException(nameof(twinCollection));
+
+            value = default;
+
+            if (!twinCollection.Contains(property))
+                return false;
+
+            // cast to object to avoid dynamic code to be generated
+            var some = (object)twinCollection[property];
+
+            // quick path for values that can be directly converted
+            if (some is JValue someJValue)
+            {
+                if (someJValue.Value is T someT)
+                {
+                    value = someT;
+                    return true;
+                }
+
+                if (someJValue.Type == JTokenType.Null)
```

Comment:
I think that this might cause a problem in the following test case (does not yet exist):

`Assert.True(reader.TryRead<ushort?>(null, out var result));`

I would have expected that parsing a `null` into a nullable type should succeed.

Reply:
I think it's still correct. You can't read anything that was not specified, hence the value you get back is correct (default) but the read was not successful. Anyway, you can't specify twins with null values as far as I know. They are removed when you try to do so.

Reply:
When going through the change, I realized that this was wrong. Since IoT hub does sanitize null values and remove them in the twins, we should not have tests that actually assume this would be the case. I removed support for that in the twin extensions completely.